### ISSUE TITLE
Can't login as local admin since user_saml 1.6.2

### ISSUE THREAD

**Issue by jagland** [nextcloud/user_saml#242]
Can't login as local admin since 1.6.2, suspect this is as a result of https://github.com/nextcloud/user_saml/pull/238

and seems to undo progress/issues raised here;

https://github.com/nextcloud/user_saml/issues/126
https://github.com/nextcloud/user_saml/pull/196

### Steps to reproduce

Attempt to login to nextcloud.
Difficult to provide steps, this worked before upgrade I was using environment-variable, and had allow multiple_user_back_ends set so that I can login as local admin.  Now it's taking me to /index.php/apps/user_saml/saml/login which triggers my external auth (turning that off just leads me to index.php/apps/user_saml/saml/notProvisioned)

from DB
```
user_saml               | type                                       | environment-variable 
user_saml               | general-allow_multiple_user_back_ends      | 1
```
### Expected behaviour
Previously I was presented with a page that had Direct Login, and what ever I had set for general-idp0_display_name
```
 user_saml               | general-idp0_display_name                  | Login with SSO
```
### Actual behaviour
Take straight to external authentication page via environment variable.

### Server configuration

**Operating system**: Ubuntu

**Web server:** Apache

**Database:** mysql  Ver 14.14 Distrib 5.7.23, for Linux (x86_64)

**PHP version:** 7.0.30-0ubuntu0.16.04.1

**Nextcloud version:** 14.0.0 Beta 3

**Where did you install Nextcloud from:** tar.bz2 from nextcloud website.

**List of activated apps:**  
```
Enabled:
  - accessibility: 1.0.1
  - activity: 2.7.0
  - bruteforcesettings: 1.1.0
  - cloud_federation_api: 0.0.1
  - comments: 1.4.0
  - dav: 1.6.0
  - federatedfilesharing: 1.4.0
  - federation: 1.4.0
  - files: 1.9.0
  - files_pdfviewer: 1.3.2
  - files_sharing: 1.6.2
  - files_texteditor: 2.6.0
  - files_trashbin: 1.4.1
  - files_versions: 1.7.1
  - files_videoplayer: 1.3.0
  - firstrunwizard: 2.3.0
  - gallery: 18.1.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.2.0
  - nextcloud_announcements: 1.3.0
  - notifications: 2.2.1
  - oauth2: 1.2.1
  - password_policy: 1.4.0
  - provisioning_api: 1.4.0
  - serverinfo: 1.4.0
  - sharebymail: 1.4.0
  - support: 1.0.0
  - survey_client: 1.2.0
  - systemtags: 1.4.0
  - theming: 1.5.0
  - twofactor_backupcodes: 1.3.0
  - updatenotification: 1.4.1
  - user_external: 0.4
  - user_ldap: 1.4.0
  - user_saml: 1.6.2
  - workflowengine: 1.4.0
Disabled:
  - admin_audit
  - encryption
  - files_external
```
**Nextcloud configuration:**

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "nextcloud.OBFUSCATED"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "http:\/\/nextcloud.OBFUSCATED",
        "dbtype": "mysql",
        "version": "14.0.0.15",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "loglevel": 2,
        "maintenance": false,
        "theme": "",
        "updater.release.channel": "beta"
    }
}
```

**Comment by brtptrs** [comment]
I can confirm this problem
My config:

        "user_saml": {
            "enabled": "yes",
            "general-allow_multiple_user_back_ends": "1",
            "general-require_provisioned_account": "0",
            "general-uid_mapping": "REMOTE_USER",
            "installed_version": "1.6.2",
            "type": "environment-variable",
            "types": "authentication"
        },

**Comment by flotpg** [comment]
Hey, got exact the same issue with current Nextcloud Build and Azure AD.
The trick ist to tick on **ADFS URL-Encodes SAML data as lowercase, and the toolkit by default uses uppercase. Enable for ADFS compatibility on signature verification** in the security settings of SSO & SAML Authentication!

**Comment by jagland** [comment]
TL;DR  try using `/index.php/login?direct=1` in your browser to auth as local admin/accounts.

Does this work for any of you @flotpg @brtptrs? as way to authenticate without using the environment variable SSO method.

I found it in my browser history as a result I guess of using the previous login page.  Doesn't appear to be clearly documented anywhere...

In my case I'm protecting `/index.php/apps/user_saml/saml/login` with Shibboleth SP, which is the configured to use my Shibboleth IdP.  Of course if you've protected more than this with your authentication module e.g. in Apache/webserver then you need to make sure it isn't protecting `/index.php/login?direct=1`

**Comment by brtptrs** [comment]
yes /index.php/login?direct=1 works for me.

Thank you

**Comment by jagland** [comment]
Looks to be just a documentation issue then...

**Comment by flotpg** [comment]
> TL;DR try using `/index.php/login?direct=1` in your browser to auth as local admin/accounts.
> 
> Does this work for any of you @flotpg @brtptrs? as a way to authenticate without using the environment variable SSO method.
> 
> I found it in my browser history as a result I guess of using the previous login page. Doesn't appear to be clearly documented anywhere...
> 
> In my case I'm protecting `/index.php/apps/user_saml/saml/login` with Shibboleth SP, which is the configured to use my Shibboleth IdP. Of course if you've protected more than this with your authentication module e.g. in Apache/webserver then you need to make sure it isn't protecting `/index.php/login?direct=1`

I can't get a local auth prompt. I'm getting forwarded to the Azure-AD Sign In Page.

**Comment by brtptrs** [comment]
> Looks to be just a documentation issue then...

I prefer the previous behaviour in the app where the user could log in using a local username+password OR could follow a link to an SSO login (CAS in our case)

**Comment by flotpg** [comment]
How can I log in with local user?

**Comment by brtptrs** [comment]
> How can I log in with local user?
Some users  can be given a normal account on the Nextcloud server, and should have the possibility to log in using the normal login page. Others get their accounts by authenticating against another database. Both should have the option to choose the appropriate option.

**Comment by flotpg** [comment]
How can I login in with local user and avoid using the configure SAML login landing page?

**Comment by jagland** [comment]
@flotpg this issue was raised related to use of the 'environment variable' option (which means you aren't using the SAML part of the plugin) and lack of admin/local account access via that. I think from what you describe you are using the built-in SAML within this plugin, so not the exact same issue?

**Comment by flotpg** [comment]
@jagland I'm not quiet sure if I can tell.
I enabled the saml and ldap as app.
`<?php
$CONFIG = array (
  'passwordsalt' => '####################################',
  'secret' => '####################################',
  'trusted_domains' =>
  array (
    0 => 'localhost',
    1 => '####################################',
    2 => '172.17.0.0/22',
    3 => '10.10.100.0/22',
    4 => '####################################',
  ),
  'datadirectory' => '/var/ncdata',
  'overwrite.cli.url' => 'https://####################################/',
  'dbtype' => 'mysql',
  'version' => '13.0.6.1',
  'dbname' => 'nextcloud_db',
  'dbhost' => 'localhost',
  'dbport' => '',
  'dbtableprefix' => 'oc_',
  'dbuser' => '####################################',
  'dbpassword' => '####################################',
  'logtimezone' => 'UTC',
  'installed' => true,
  'instanceid' => '####################################',
  'appstore.experimental.enabled' => 'true',
  'mail_smtpmode' => 'smtp',
  'mail_smtpauth' => 1,
  'mail_smtpport' => '587',
  'mail_smtphost' => 'smtp.office365.com',
  'mail_smtpauthtype' => 'LOGIN',
  'mail_from_address' => 'smtp',
  'mail_domain' => '####################################',
  'mail_smtpsecure' => 'tls',
  'mail_smtpname' => '####################################',
  'mail_smtppassword' => '####################################',
  'preview_libreoffice_path' => '/usr/bin/libreoffice',
  'memcache.local' => '\\OC\\Memcache\\Redis',
  'filelocking.enabled' => true,
  'memcache.distributed' => '\\OC\\Memcache\\Redis',
  'memcache.locking' => '\\OC\\Memcache\\Redis',
  'redis' =>
  array (
    'host' => '/var/run/redis/redis.sock',
    'port' => 0,
    'timeout' => 0,
    'dbindex' => 0,
    'password' => '####################################',
  ),
  'htaccess.RewriteBase' => '/',
  'mysql.utf8mb4' => true,
  'maintenance' => false,
  'theme' => '',
  'loglevel' => 2,
  'ldapIgnoreNamingRules' => false,
  'ldapProviderFactory' => '\\OCA\\User_LDAP\\LDAPProviderFactory',
  'ldapUserCleanupInterval' => 10,
);`

<img width="604" alt="saml" src="https://user-images.githubusercontent.com/20798074/47006017-48d21700-d135-11e8-975b-db46912f5057.png">
<img width="831" alt="apps" src="https://user-images.githubusercontent.com/20798074/47006018-496aad80-d135-11e8-97d8-aaface3b823a.png">

**Comment by flotpg** [comment]
As I can see, the issue was i was on v13.
Now I have issues that the userPrincipalNames in AD sometimes have uppercase characters. These users are synched to Azure AD and it looks like they are normalized there to all lowercase resulting in double created users in OC

**Comment by jagland** [comment]
@flotpg  You are using the built-in SAML part, not the 'environment-variable' option which disables that, the issue I raised related here to where the environment-variable option (my SAML implementation - Shibboleth SP runs outside of Nextcloud), so your issue is different/off topic.

Maybe you've found a bug with the interaction between the user_saml plugin and ldap plugin?   Assuming your attribute is trying to be matched between them?   Sorry I can't help anymore I'm #just-another-user!
